### ISSUE TITLE
RFC - Simplify build process

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone#749]
EDIT: here is the status of the 8 PRs described below:

- [x] PR1
- [x] PR2
- [ ] PR3
- [ ] PR4
- [ ] PR5
- [ ] PR6
- [ ] PR7
- [ ] PR8

---

I want to change how we do builds. Today we do something like this:

* generate build script, save to disk
* generate ssh keys, save to disk
* generate dockerfile with env vars, save to disk
* create image
* start container that runs build script in container
* delete container
* delete build script

I would like to simplify:

* start container, injecting the build script as a command line arg to the container
* delete container

To do this we can pass the build script into the Docker like this:

```
docker run {image} /bin/bash -c "echo {script} | base64 --decode | tee run.sh | sh run.sh && rm run.sh"`
```

The build script is base64 encoded in the command line argument. The build script is then removed once the build is complete.

This is a big change, and since there have already been a lot of big changes, I would like to split this into a series of PRs to minimize impact to the stability of the project. Having multiple PRs will let us test each change in isolation and make it easier to troubleshoot issues. It will also make code reviews easier :)

### PR 1
the build script should be responsible for writing the ssh key to the filesystem instead of ADD in the Dockerfile. See the following code:
https://github.com/drone/drone/blob/master/shared/build/build.go#L561

Below is an example of what the bash function might look like.  Note that Docker may set $HOME=/ by default. We may need to query the user home directory. I'm not entirely sure though ... needs testing

```sh
[ -d $HOME/.ssh ] && mkdir -p $HOME/.ssh
chmod 700 $HOME/.ssh
cat %q > $HOME/.ssh/id_rsa
chmod 600 $HOME/.ssh/id_rsa
```

### PR 2
the build script should export all environment variables, instead of using the Dockerfile. See:
https://github.com/drone/drone/blob/master/shared/build/build.go#L489

### PR 3
the build script should no longer user `/var/cache/drone`. It should instead use the default `WORKINGDIR`. We will also need to account for `$GOPATH`. We also need to account for the fact that the git clone path can be specified in the `.drone.yml`

here is what I'm thinking:

```sh
export DRONEPATH = $HOME # or git clone directory in yaml, if set
[ -v $GOPATH ] && DRONEPATH = $GOPATH/src
cd $DRONEPATH

# ... rest of build script. git clone, etc
```

### PR 4
the build script should use the default USER. It should never attempt to set or override the USER

### PR 5
I feel like we no longer need to the `proxy.sh` command. We should be able to include a simple bash function in the shell script that automatically checks for environment variables with the appropriate suffix (ie `*_TCP_PORT`) and handle it dynamically. See:
https://github.com/drone/drone/blob/master/shared/build/build.go#L541

Note that today, the `proxy.sh` script is stored in `/etc/drone.d`. Others have started using this directory to initialize their environment. We should let people know it is being deprecated prior to removing it:

```sh
[ -d /etc/drone.d ] echo ‘WARNING: /etc/drone.d IS DEPRECATED, THIS WILL BE REMOVE IN NEXT RELEASE'
```

## PR 6
Some OS level configuration. We need to ensure we disable strict host key checking for git. Note that this should take into account the fact that the user may not be root (and use sudo)

```
echo 'StrictHostKeyChecking no' > /root/.ssh/config
```

We also need to ensure `apt-get` can run without `-y` by adding the below text file to `/etc/apt/apt.conf.d/90forceyes` with the correct OS permissions. This should also account for the fact that the user may not be root (and use sudo). It should account for the fact that sudo may not be included in the image. It should also account for the fact that the image may not be debian-based. None of these scenarios should fail the build.

```
APT::Get::Assume-Yes "true";APT::Get::force-yes "true";
```

### PR 7
Now that everything is handled in the build script we can alter the Docker entrypoint and command to something like this:

```
/bin/bash -c "echo %s | base64 --decode | tee run.sh | sh run.sh && rm run.sh"
```

Note that we can remove the code that creates the build image and deletes the build image.

### PR 8
While we're at it, let's get rid of the `docker` package in Drone and start using Docker client created by the Docker guys:
https://github.com/samalba/dockerclient

---

### Question 1
One thing I don't have figured out is handling local builds, using the `drone build` command. I think we should slightly alter the approach and mount the code as a volume, but I'm not sure this will work on a non-linux machine running boot2docker. Can boot2docker mount volumes on the mac host?

https://github.com/drone/drone/blob/master/shared/build/build.go#L154

**Comment by ianmacl** [comment]
echo 'StrictHostKeyChecking no' > /root/.ssh/config - is this really a good idea?  Perhaps it is better to have the option of adding known hosts to a config file or something?  I assume this is to get around problems with ssh not recognizing the host.

Maybe it's fine because we're containerized, but strict host key checking off makes me nervous.

**Comment by bradrydzewski** [comment]
This is how I envision us handling PR #5:

```sh
for v in $(compgen -A variable | grep "_TCP_ADDR");
do
        ADDR=${!v}
        PORT=${v/ADDR/PORT}
        PORT=${!PORT}
        socat TCP-LISTEN:$PORT,fork TCP:$ADDR:$PORT &
done
```

The `build.sh` file can determine what to proxy with `socat` (or polyfill script in #869) by looking at the environment variables, instead of us having to explicitly inspect the container for exposed ports in our Go code, generating shell commands, etc

**Comment by dave-tucker** [comment]
@bradrydzewski re:boot2docker, it *can* mount remote volumes now but [there are restrictions](https://github.com/boot2docker/boot2docker#virtualbox-guest-additions).

I've been using it to build Docker and other projects on OSX and it works well for me

`docker run --rm -v $PWD:/usr/src/application ...`

**Comment by bradrydzewski** [comment]
abandoned in favor up new yaml and plugin system
